### PR TITLE
DataLinks: Fix blur issues

### LINKED ISSUE #19868: DataLinks: in 6.4.3, the url field still doesn't release focus despite #19804
**What happened**:

I was creating a datalink (well, trying to create) in 6.4.3 and after I started entering the URL, I could only add one character to the Title field before focus switched to the URL field. This is despite pull #19804 theoretically fixing this issue.

**What you expected to happen**:

After starting to enter the URL, I could switch back to the Title field and enter (and update) a title there.

**How to reproduce it (as minimally and precisely as possible)**:

Start adding a data link. Enter text in the Title field; it works fine. Switch focus to the URL field and type one character. Now switch focus back to the Title field and try to type more characters.

**Environment**:
- Grafana version: 6.4.3
- Data source type & version: Prometheus 2.13.0
- OS Grafana is installed on: Ubuntu 18.04 (64-bit)
- User OS & Browser: Fedora Linux 30 (64-bit), Official Mozilla.org Firefox 69.0.3 
- Grafana plugins: flant-statusmap-panel  natel-discrete-panel

### PR DESCRIPTION
**Pull request by aocenas**
Closes: https://github.com/grafana/grafana/issues/19868
Related to: https://github.com/grafana/grafana/pull/19804 which did not fix all cases.

The main issue was that we updated parent state in the `onBlur` handler. This triggered something like this:
- in onBlur we called parents onChange with new string state
- we called `next()` which would eventually trigger local `onChange` with new state of the Editor which also contains the `selection` and focus/unfocus state.
- in the meantime the change of state in parent triggered rerender which was hard to prevent even with memoization.
- this rerender was done with the old editor value where the selection was still focused essentially capturing the focus again.
- The change of state to unfocused was canceled because the editor got new value, so the blurred state was never propagated to our local `onChange` and local editor state (I think, not 100% sure on this).

So the fix is basically not to do anything in `onBlur` itself but wait until the focus changed is propagated into the local state, and only after that is rendered the parents `onChange` is called propagating the new value upstream.

### CODE REVIEW

File: packages/grafana-ui/src/components/DataLinks/DataLinkInput.tsx [hunk 1]
```diff
@@ -86,18 +88,19 @@ export const DataLinkInput: React.FC<DataLinkInputProps> = memo(({ value, onChan
     }
   }, []);
 
+  useEffect(() => {
```
Comment (davkal) [inline packages/grafana-ui/src/components/DataLinks/DataLinkInput.tsx]: Works, but a bit weird that we have to track blur like this.

**Review by davkal — COMMENTED** [review]
The outer function's `onChange` argument is still changed every time the caller renders, making the `memo` useless.

Is there a way to test the editor (title + url fields)?

Otherwise this can go in.

File: packages/grafana-ui/src/components/DataLinks/DataLinkInput.tsx (hunk 1 above)
Comment (aocenas) [reply packages/grafana-ui/src/components/DataLinks/DataLinkInput.tsx]: Well the main point is we mix imperative editor control with controlling the value of the Editor. We should probably stick to just one. Like here we probably do not need to have Value in the local state at all and just use editor commands instead (this seems like better documented by slate).

Other way would be to go only by state changes but then we would need a way to easily trigger changes with some help utils (like `setFocused(value)` and `isFocused(value)`) but slate does not seem to have such.

So in essence this is "React" way to do this, handlers and commands due to their implementation is more imperative slate way to do it. For example the onBlur handler is changing the state of the Editor, "React" way would be to do nothing and let the component change the local Value and pass it to the Editor with `focus=false`.

**Comment by aocenas** [comment]
> Is there a way to test the editor (title + url fields)?

Will try to add tests, but not sure how to test this particular issue. Not sure if it is possible to simulate focus/blur behaviour in enzyme (would think not). Also last time I had problem mounting Slate in enzyme in general so will have to take a look at that.

**Comment by marefr** [comment]
Should this issue/PR be included in next patch release? Missing a milestone and possibly cherry-pick needed.

**Comment by aocenas** [comment]
@marefr I think it should. It is missing test which I did not have time to implement so far as it needs some research but if we going to do a release soon we can merge this and figure out testing separately.

**Comment by marefr** [comment]
Okay :+1: . No date for release right now, but you can add milestone so that we can see that this is something to be included or that we want to include.

**Review by davkal — APPROVED** [review]
Still good to merge.

**Pull request merged**